### PR TITLE
Patch 1

### PR DESCRIPTION
**Pull request by krebeck**
This pull request adds functionality improved on and referenced in this discussion:

https://groups.google.com/forum/#!topic/islandora/EwWeXNJ-dlk

This code adds configuration for a simple catalog search text to appear beneath the Google Scholar search text.  It is configurable in much the same way and also requires changes to utilities.inc in this directory.

It provides a way to query a discovery layer without needing an outside block or additional modules.

# How should this be tested?

This can be tested by adding the required configurations from a known good discovery layer and then testing from a citation object.

# Additional Notes:
This should not require any substantive change to the documentation.  It's pretty self-explanatory.  I improved my first draft so that you can also customize the search parameters, but it provides a good set of defaults for those using the popular Summon platform - that should cover ~30-50% of academic libraries.  So it may not work out of the box for everyone, but is easily configurable for everyone.

### CODE REVIEW

**Comment by bryjbrown** [comment]
I loaded this up in a test VM and it works as I expected. Turned off by default which is good. Not sure whats going one with Travis though, I see the error is from:
```
The command "drush coder-review --reviews=production,security,style,i18n,potx,sniffer --no-empty islandora_scholar" exited with 1.
```
@DiegoPino, is this related to the previous Travis issues we'd been having?

File: includes/admin.form.inc
Comment (bryjbrown) [inline includes/admin.form.inc]: Curious as to why you want to replace this title finding xpath query with a less specific one?

File: includes/utilities.inc
Comment (bryjbrown) [inline includes/utilities.inc]: Same issue as previous comment, why the xpath replacement?

**Comment by DiegoPino** [comment]
@bryjbrown no, seems like different hiccup from Travis. Restarting now because can't find the logs, but could be also long lines and PHPCS? Also, @krebeck, this needs a JIRA ticket.  And please excuse my ignorance, But are you or your institution covered by our CLA? Thanks for this contribution!

**Comment by DiegoPino** [comment]
Ok, now i got the logs. Drupal Coding standards problem
https://travis-ci.org/Islandora/islandora_scholar/jobs/226160920#L959
Please run `phpcs --standard=Drupal your files...` on your files before committing to be sure they pass. `phpcbf` also fixes many of the issues automatically

**Comment by manez** [comment]
@krebeck I think you're at WRLC? If so, as @DiegoPino notes, we'll need to have a CLA for you or your institution in order to merge this contribution. You can read more (and get the forms) [here](https://github.com/Islandora/islandora/wiki/Contributor-License-Agreements) or ping me at manez@islandora.ca if you need more info

Thanks!

**Comment by manez** [comment]
iCLA now on file for @krebeck

**Comment by DonRichards** [comment]
OK, but I'm still wondering why is this pointing to Proquest? I didn't see anywhere in the description why Proquest options are essential for this to "_adds configuration for a simple catalog search text to appear beneath the Google Scholar search text_".

**Comment by bryjbrown** [comment]
@DonRichards Can you link to the specific lines you are referring to? I don't see anything about ProQuest in the PR.

**Comment by krebeck** [comment]
The default catalog link (line 90) does point to American University's Summon instance as an example.  The Summon discovery service is a ProQuest product.  But I wanted to put something real in that field to demonstrate how it should work.  If it would be better to anonymize with something like "your.service.search" or whatever, that's fine with me.

**Comment by DonRichards** [comment]
@bryjbrown [includes/utilities.inc](https://github.com/Islandora/islandora_scholar/pull/265/files#diff-2d69368f59ba81d5635cd8274dd66090R82). @krebeck pointing to a 3rd party is unusual for Islandora from my experience with the community. I think pointing to your.service.search would work better but I was wondering why the [fallback](https://github.com/Islandora/islandora_scholar/pull/265/files#diff-2d69368f59ba81d5635cd8274dd66090R82) was them as well.

**Comment by bryjbrown** [comment]
@krebeck I'm going to have to agree with @DonRichards, while having the URL there certainly makes it easier to test, it could cause problems if the URL changes (which eResource services are prone to doing). My recommendation would be to set the default variable to an empty string and check to make sure that value is NOT an empty string on form submission (easy to do) to make sure folks actually enter something if they are going to enable the service. Entering something like "http://your.search.service" could work too, but I think having an empty string with good help text would be the better choice.

Also @krebeck don't forget to address the other code review issues too (changing the default search xpath) as I'm interested in knowing why you changed these.

File: includes/admin.form.inc
Comment (bryjbrown) [inline includes/admin.form.inc]: As @DonRichards points out, we shouldn't default to a third party service. Either put in an obviously fake URL like "your.search.service" or an empty string and check to make sure that this default value is not set on form submission.

**Comment by DonRichards** [comment]
Travis failing: 
```shell
Drush command terminated abnormally due to an unrecoverable error.   [error]

The command "drush coder-review --reviews=production,security,style,i18n,potx,sniffer --no-empty islandora_scholar" exited with 1.

$ drush coder-review --reviews=production,security,style,i18n,potx,sniffer --no-empty bibutils

Release standards, Drupal Security Checks, Drupal Coding Standards, Internationalization, Interface text translatability, Drupal CodeSniffer
```

I  have no reviewed the error yet but thought it would save everyone the effort of navigating to Travis.

**Comment by DiegoPino** [comment]
Ok Folks. I hope there is agreement on the changes. In the meanwhile, this has a lot of drupal coding issues, most very harmless that can be solved via `phpcbf`.

```Shell
sites/all/modules/islandora_scholar/includes/admin.form.inc:
 +91: [normal] Use an indent of 2 spaces, with no tabs
 +91: [minor] Spaces must be used to indent lines; tabs are not allowed
 +92: [normal] Use an indent of 2 spaces, with no tabs
 +92: [minor] Spaces must be used to indent lines; tabs are not allowed
 +93: [minor] Avoid backslash escaping in translatable strings when possible, use "" quotes instead
 +103: [minor] Avoid backslash escaping in translatable strings when possible, use "" quotes instead
sites/all/modules/islandora_scholar/includes/utilities.inc:
 +80: [minor] Inline comments must end in full-stops, exclamation marks, or question marks
 +83: [minor] Inline comments must end in full-stops, exclamation marks, or question marks
````

@krebeck could you fix those? Also. Since this options are kinda specific and not everybody will have those, do you all think we need more documentation and examples? i would hate to see this functionality become an obscure part of out stack.

thansk!!

**Comment by DiegoPino** [comment]
@krebeck any updates on this? I still see Travis-CI failing because of coding standards. Please let me or any other committer know if you need help with that. Also, just for the sake of documenting this new features, could you open a JIRA ticket? (same, if you need help, happy to assist) Thanks!!

**Comment by krebeck** [comment]
@DiegoPino I have made all of the edits referenced in the php coding issues above.  It should be clean now.  If you could assist with the JIRA portion, that would be a huge help.

**Comment by bryjbrown** [comment]
@krebeck Please don't forget to address the code review issues, specifically your changing of `islandora_scholar_google_scholar_default_search_xpath`.

**Comment by bryjbrown** [comment]
Pinging @krebeck about the code review as I'd like to get this merged before the 7.x-1.11 code freeze.

If any @Islandora/7-x-1-x-committers have opinions about the XPaths used in this PR (see code review) vs the ones currently in Scholar, please chime in. If I don't hear from anyone before the code freeze, I would vote for changing those XPaths back to `//mods:mods[1]/mods:titleInfo/mods:title` in this PR and then merging.

My reasoning:
`//mods:mods[1]/mods:titleInfo/mods:title` is the current default value for `islandora_scholar_google_scholar_default_search_xpath` as it very specifically targets the `mods:title` as found in a `mods:titleInfo` element that is the direct child of the `mods:mods` wrapper element. 

`//mods:title`, which may work properly in some cases, could get confused by a `mods:title` element literally anywhere else. For example, if a MODS record for a journal article referenced the title of the journal as a `mods:relatedItem`, it would do so inside of a `mods:title` element. Here's an example of what that would look like in practice:

```
<mods:mods>
    <mods:titleInfo>
        <mods:title>This is the title of the article</mods:title>
    <mods:titleInfo>
    <mods:relatedItem type="host">
         <mods:titleInfo>
             <mods:title>This is the title of the journal in which the article appears</mods:title>
         <mods:titleInfo>
    </mods:relatedItem>
</mods:mods>
```

In this situation, the `//mods:title` XPath could just as easily match on the title of the journal as the title of the actual article being described.

This PR has been sitting around for a while and I would really like to get it merged, but I need for the code review to be addressed before I can do so.

**Comment by DiegoPino** [comment]
@bryjbrown I agree with you on the ```mods:title``` issue. This pull has still a JIRA ticket missing, any @Islandora/7-x-1-x-committers who has the energy to do so? @bryjbrown code freeze already announced?

**Comment by bryjbrown** [comment]
@DiegoPino 
> code freeze already announced?

I haven't heard anything about it, but I'm trying to be a bit more proactive on the maintainer front than I was in previous releases when the code freeze snuck up on me while I was napping :)

**Comment by whikloj** [comment]
I'd also note that the added variables need to be added to the `islandora_scholar_uninstall()` to ensure they are cleaned up on un-install.

**Comment by whikloj** [comment]
@bryjbrown @DiegoPino @DonRichards 

[Here is](https://github.com/Islandora/islandora_scholar/compare/7.x...uml-digitalinitiatives:krebeck-patch-1-clean?expand=1) my clean-up of @krebeck's PR (compared against 7.x).

If it resolves your concerns I'll open a PR against their PR and see if we can't move this through.

**Review by bryjbrown — CHANGES_REQUESTED** [review]
So apparently I made comments on the code in question, but never clicked "Submit review" so only I could see the code comments in question. MY BAD. This is the phantom code review I was referring to in the 03/08 Islandora Committers Call, during which I ALSO could not see the code review comments because I was logged out. ¯\_(ツ)_/¯

**Comment by bryjbrown** [comment]
@whikloj Nice job! This adds a lot of good stuff, like the default URL description and the variable uninstallations. One of the things in your PR is actually the thing I have a problem with, which is the resetting of `islandora_scholar_google_scholar_default_search_xpath`. It turns out I never formally submitted the code review that I was referring to so only I could see it, but now that its publicly visible please check it out and LMK what you think.

**Comment by whikloj** [comment]
@bryjbrown Yeah I figured it out from comments. 

So I [reverted the change](https://github.com/Islandora/islandora_scholar/compare/7.x...uml-digitalinitiatives:krebeck-patch-1-clean?expand=1#diff-0cdd19abe14e51fe9e7f061ec72c86d3R75) to `islandora_scholar_google_scholar_default_search_xpath` (back to `//mods:mods[1]/mods:titleInfo/mods:title`), but I left the new 
`islandora_scholar_library_catalog_default_search_xpath` variable with the default `//mods:title`.

I can amend that to the same as `islandora_scholar_google_scholar_default_search_xpath`. That didn't seem to be a comment on the original PR so I tried not to just start changing stuff.

**Comment by whikloj** [comment]
@bryjbrown Actually this might be an easier way to see it, this is the one commit I added on to this PR. It shows what I changed.
https://github.com/Islandora/islandora_scholar/commit/b97610331d1311fb1dda9f4129d9699c2cfe59b1

**Comment by bryjbrown** [comment]
@whikloj I see now. I don't have any issues with merging this so long as we get it back to `variable_get('islandora_scholar_google_scholar_default_search_xpath', '//mods:title')` in your PR. Thats the only thing left that I have any issue with.

Once we get to that step, you may need to guide me through the process of merging a PR into a PR so we can merge the PR :sweat_smile:

**Comment by whikloj** [comment]
@bryjbrown just to clarify, this is correct?
`variable_get('islandora_scholar_google_scholar_default_search_xpath', '//mods:mods[1]/mods:titleInfo/mods:title')`

**Comment by bryjbrown** [comment]
@whikloj AHHHHH yes you are correct. It is the longer more specific xpath that is correct, not the shorter one. I'm sorry, its been a very long day.

**Comment by whikloj** [comment]
I have opened a PR against @krebeck's PR (https://github.com/krebeck/islandora_scholar/pull/1), once it is merged it will automatically apply to this PR and you should see the changes here. So you won't have to merge a PR on a PR. 😄

**Comment by whikloj** [comment]
I also made a Jira ticket for this. https://jira.duraspace.org/browse/ISLANDORA-2174

**Comment by bryjbrown** [comment]
@whikloj So that PR against @krebeck's fork can only be merged by him. 

@krebeck, care to pull the trigger on this?

**Review by bryjbrown — APPROVED** [review]

**Comment by bryjbrown** [comment]
Alrighty, it looks like @krebeck merged @whikloj's PR into his local branch so now this parent PR seems good to go for me.

@DiegoPino still not sure what the protocol is for merging, do I need to wait 24 hours for final comments, or can I pull the trigger on it now?

**Comment by krebeck** [comment]
Thanks to all for your help.  I'm obviously a little out of my depth here
especially with the administrative aspects of PRs.

Aaron Krebeck
Digital Services Librarian
Washington Research Library Consortium
301-390-2040

**Comment by whikloj** [comment]
@krebeck no worries, your first PR is all about a solution you wanted to share.

Its not until you become old and grizzled like me that you start shouting about default values. 😄

**Comment by DiegoPino** [comment]
@bryjbrown soft rule is 24hours after approval if nobody opposed to the approval or asked for changes post approval of course.  Some say after the last tested change, some say after concerns where addressed. All say after someone approved.
But in this case, if all involved (sorry this excludes the ones that committed!) are Ok with merging because of some higher reason (code freeze, server on fire, in love, etc), then yes. If we can wait 22 hours in this case, even better.

**Comment by bryjbrown** [comment]
Okay, I'll merge the code and close the JIRA ticket tomorrow then. Hooray!

**Comment by whikloj** [comment]
Welcome to the other side @krebeck.

**Pull request merged**
